# Introducción a la manipulación de datos con pandas

<img style="float: right; margin: 0px 0px 15px 15px;" src="https://upload.wikimedia.org/wikipedia/commons/e/ed/Pandas_logo.svg" width="400px" height="400px" />

> El verdadero poder de python viene con sus librerías. Ya en cursos pasados han aprendido varios aspectos relativos a las librerías ([NumPy](https://www.numpy.org/) y [matplotlib](https://matplotlib.org/)). En estos enlaces pueden encontrar la documentación oficial para revisar qué más pueden hacer con estas librerías. También el `help` puede ser un buen punto de partida para conocer qué más cosas pueden realizar.

> Hoy nos dedicaremos a estudiar la librería de Python para análisis de datos: **Pandas**. Durante el transcurso de la asignatura haremos uso extensivo de datos históricos de precios y rendimientos de activos financieros, por tanto es importante tener un manejo básico de pandas.

Referencias:
- https://pandas.pydata.org/
- https://towardsdatascience.com/data-science-with-python-intro-to-loading-and-subsetting-data-with-pandas-9f26895ddd7f
___

# 0. Motivación

Normalmente trabajamos con grandes cantidades de datos (precios). 

Los datos que debemos cargar pueden guardarse de muchas maneras distintas: archivos CSV, archivos de Excel, etcétera. Incluso, los datos pueden estar disponibles a través de servicios web. 

Para trabajar con datos, se hace necesario representarlos en una estructura tabular (cualquier cosa con forma de tabla con filas y columnas).

En algunos casos, los datos ya están en forma tabular y es más fácil cargarlos. En otros, debemos trabajar con datos no estructurados o que no están organizados de una manera determinada (texto plano, imágenes, audio, etcétera).

En esta clase nos vamos a concentrar en cargar datos desde archivos CSV (valores separados por coma).

## Pandas

Pandas es una librería de código abierto para el lenguaje de programación Python, desarrollada por Wes McKinney. Es una librería muy eficiente y proporciona estructuras de datos y herramientas de análisis muy fáciles de usar.

Como las librerías que hemos visto antes, Pandas viene instalado por defecto con Anaconda, así que lo único que tenemos que hacer para empezar a trabajar con ella es importarla. La comunidad utiliza normalmente la abreviación pd para referirse a pandas:

In [1]:
# Importar pandas
import pandas as pd

Los **pd.DataFrames** son los objetos por excelencia de pandas para manipular datos. Son eficientes y rápidos. Son la estructura de datos donde pandas carga los diferentes formatos de datos: cuando nuestros datos están limpios y estructurados, cada fila representa una observación, y cada columna una variable o característica. Tanto las filas como las columnas pueden tener etiquetas.

En esta clase vamos a trabajar con datos pertenecientes a ejemplos de la librería featuretools [en esta página](https://docs.featuretools.com/loading_data/using_entitysets.html). Sin embargo para entender cómo es que se cargan estos datos, estos están disponibles en la carpeta data en formato CSV.

En esta y la siguiente clase, además de importar datos, aprenderemos a:
- seleccionar subconjuntos de datos;
- filtrar variables por categorías;
- relacionar tablas con datos complementarios;
- entre otros.

Comenzamos:

___
# 1. Importando datos

En la carpeta "data" tenemos los archivos "customers_data.csv", "sessions_data.csv", "transactions_data.csv" y "products_data.csv", cada uno de ellos conteniendo información relevante acerca de transacciones de cierta empresa.

El primer paso para comenzar a trabajar con datos es importarlos. Lo podemos hacer con la función `pd.read_csv()`:

In [2]:
# Ayuda en la función pd.read_csv()
data = pd.read_csv("data/customers_data.csv")
data.head()

Unnamed: 0  customer_id  zip_code            join_date date_of_birth
0           0            1     60091  2011-04-17 10:48:33    1994-07-18
1           1            2     13244  2012-04-15 23:31:04    1986-08-18
2           2            3     13244  2011-08-13 15:42:34    2003-11-21
3           3            4     60091  2011-04-08 20:08:14    2006-08-15
4           4            5     60091  2010-07-17 05:27:50    1984-07-28

Muy bien, concentrémonos en los datos de cliente (comenzaremos con estos datos pequeños para ilustrar las funcionalidades, luego iremos añadiendo tamaño):

In [4]:
# Importar customers_data.csv


customer_id  zip_code            join_date date_of_birth
0            1     60091  2011-04-17 10:48:33    1994-07-18
1            2     13244  2012-04-15 23:31:04    1986-08-18
2            3     13244  2011-08-13 15:42:34    2003-11-21
3            4     60091  2011-04-08 20:08:14    2006-08-15
4            5     60091  2010-07-17 05:27:50    1984-07-28

Bien, ya tenemos los datos cargados, sin embargo, se ve algo raro. 

¿Qué es esta columna "Unnamed: 0"? (abrir el CSV)

Para especificar que esta columna corresponde al índice, podemos usar el argumento `index_col`.

In [5]:
data = pd.read_csv("data/customers_data.csv", index_col = [0] )
data.head()

customer_id  zip_code            join_date date_of_birth
0            1     60091  2011-04-17 10:48:33    1994-07-18
1            2     13244  2012-04-15 23:31:04    1986-08-18
2            3     13244  2011-08-13 15:42:34    2003-11-21
3            4     60091  2011-04-08 20:08:14    2006-08-15
4            5     60091  2010-07-17 05:27:50    1984-07-28

In [4]:
# Importar customers_data.csv, haciendo uso del argumento index_col


In [6]:
# Tipo de objeto con que importa pandas los archivos
print(type(data))

<class 'pandas.core.frame.DataFrame'>


Ahora nuestro DataFrame luce bien. Sin embargo, podríamos querer indizar nuestro DataFrame directamente por el id de cliente, utilizando el método `set_index()`:

In [10]:
# Usar el método set_index para indizar por el id de cliente
data_0 = data.set_index("customer_id")
data_0.head()

zip_code            join_date date_of_birth
customer_id                                             
1               60091  2011-04-17 10:48:33    1994-07-18
2               13244  2012-04-15 23:31:04    1986-08-18
3               13244  2011-08-13 15:42:34    2003-11-21
4               60091  2011-04-08 20:08:14    2006-08-15
5               60091  2010-07-17 05:27:50    1984-07-28

In [9]:
# Revisar que paso con la tabla original
data.head()

customer_id  zip_code            join_date date_of_birth
0            1     60091  2011-04-17 10:48:33    1994-07-18
1            2     13244  2012-04-15 23:31:04    1986-08-18
2            3     13244  2011-08-13 15:42:34    2003-11-21
3            4     60091  2011-04-08 20:08:14    2006-08-15
4            5     60091  2010-07-17 05:27:50    1984-07-28

Podemos ver que pandas no convierte **customer_id** en el índice permanente del DataFrame, a menos que se lo especifiquemos. Para especificarlo, debemos usar el argumento `inplace`

In [11]:
# Usar el método set_index para indizar por el id de cliente, usando el argumento inplace
data.set_index("customer_id", inplace = True)

In [12]:
# Revisar que paso con la tabla original
data

zip_code            join_date date_of_birth
customer_id                                             
1               60091  2011-04-17 10:48:33    1994-07-18
2               13244  2012-04-15 23:31:04    1986-08-18
3               13244  2011-08-13 15:42:34    2003-11-21
4               60091  2011-04-08 20:08:14    2006-08-15
5               60091  2010-07-17 05:27:50    1984-07-28

___
# 2. Indización y selección de datos

Hay muchas formas de las cuales podemos seleccionar datos de DataFrames. Veremos, de acuerdo al artículo al final de este documento, la forma basada en corchetes ([]) y en los métodos `loc()` y `iloc()`.

Con los corchetes, podemos seleccionar ciertas filas, o bien, ciertas columnas. 

Para una selección de filas, podemos usar el indizado como en las listas: [start_index:end_index:step], recordando que el `end_index` no es inclusivo.

Por ejemplo, seleccionar los clientes en las primeras dos filas:

In [13]:
data.iloc[:2]

zip_code            join_date date_of_birth
customer_id                                             
1               60091  2011-04-17 10:48:33    1994-07-18
2               13244  2012-04-15 23:31:04    1986-08-18

Ahora, seleccionar del tercer cliente en adelante:

In [21]:
data.iloc[2:]

zip_code            join_date date_of_birth
customer_id                                             
3               13244  2011-08-13 15:42:34    2003-11-21
4               60091  2011-04-08 20:08:14    2006-08-15
5               60091  2010-07-17 05:27:50    1984-07-28

Finalmente, seleccionar los clientes en las filas impares:

In [20]:
data.loc[::2]

zip_code            join_date date_of_birth
customer_id                                             
1               60091  2011-04-17 10:48:33    1994-07-18
3               13244  2011-08-13 15:42:34    2003-11-21
5               60091  2010-07-17 05:27:50    1984-07-28

Similarmente, para una selección de columnas, podemos usar una lista con los nombres de las columnas requeridas. Si pasamos solo una columna como un string en lugar de una lista, el resultado será un objeto **pd.Series**:

In [25]:
# Seleccionar la columna date_of_birth como Series
data["date_of_birth"]


customer_id
1    1994-07-18
2    1986-08-18
3    2003-11-21
4    2006-08-15
5    1984-07-28
Name: date_of_birth, dtype: object

In [26]:
# Tipo de objeto pd.Series
print(type(data["date_of_birth"]))

<class 'pandas.core.series.Series'>


Las Series de pandas son arreglos unidimensionales que pueden ser etiquetados. En algunos casos, querremos seleccionar solo una columna pero mantener esta selección en un DataFrame. En este caso podemos pasar una lista con un solo elemento:

In [27]:
# Seleccionar la columna date_of_birth como DataFrame
data[["date_of_birth"]]

date_of_birth
customer_id              
1              1994-07-18
2              1986-08-18
3              2003-11-21
4              2006-08-15
5              1984-07-28

In [28]:
# Tipo de objeto pd.DataFrame
print(type(data[["date_of_birth"]]))

<class 'pandas.core.frame.DataFrame'>


Finalmente, seleccionamos dos columnas:

In [32]:
# Seleccionar las columnas date_of_birth y zip_code
data[["date_of_birth","zip_code"]]

date_of_birth  zip_code
customer_id                        
1              1994-07-18     60091
2              1986-08-18     13244
3              2003-11-21     13244
4              2006-08-15     60091
5              1984-07-28     60091

Muy bien, ya vimos que los corchetes son útiles. También existen los poderosos métodos `loc` y `iloc`, que nos dan el poder de seleccionar ambos a la vez: columnas y filas.

¿En qué se diferencian?

- El método `loc` nos permite seleccionar filas y columnas de nuestros datos basados en etoquetas. Primero, se deben especificar las etiquetas de las filas, y luego las de las columnas.

- El método `lioc` nos permite hacer lo mismo pero basado en índices enteros de nuestro DataFrame (como si fueran matrices).

Como antes, si queremos seleccionar todas las filas, o columnas, simplemente escribimos `:` en el lugar adecuado.

Mejor con ejemplos:

Para un mejor entendimiento de esta parte, reetiquetaremos la numeración de clientes con una "numeración" alfabética. Es decir: 1-A, 2-B, ..., 5-E.

In [ ]:
# Resetear índice en el lugar
data.reset_index(inplace = True)
data.index = ["A" , "B" , "C" , "D", "E"]

In [37]:
# Revisar la tabla
data

level_0  index  customer_id  zip_code            join_date date_of_birth
A        0      0            1     60091  2011-04-17 10:48:33    1994-07-18
B        1      1            2     13244  2012-04-15 23:31:04    1986-08-18
C        2      2            3     13244  2011-08-13 15:42:34    2003-11-21
D        3      3            4     60091  2011-04-08 20:08:14    2006-08-15
E        4      4            5     60091  2010-07-17 05:27:50    1984-07-28

In [17]:
# Reasignar índice alfabético


Ahora sí.

Seleccionemos el primer cliente con ambos métodos:

In [41]:
# Primer cliente con loc
data.loc["A"]

level_0                            0
index                              0
customer_id                        1
zip_code                       60091
join_date        2011-04-17 10:48:33
date_of_birth             1994-07-18
Name: A, dtype: object

In [42]:
# Primer cliente con iloc
data.iloc[0]

level_0                            0
index                              0
customer_id                        1
zip_code                       60091
join_date        2011-04-17 10:48:33
date_of_birth             1994-07-18
Name: A, dtype: object

Nos dan series, ¿y si queremos DataFrames?

In [43]:
# Primer cliente con loc en DataFrame
data.loc[["A"]]

level_0  index  customer_id  zip_code            join_date date_of_birth
A        0      0            1     60091  2011-04-17 10:48:33    1994-07-18

In [44]:
# Primer cliente con iloc en DataFrame
data.iloc[[0]]

level_0  index  customer_id  zip_code            join_date date_of_birth
A        0      0            1     60091  2011-04-17 10:48:33    1994-07-18

Ahora, seleccionemos los clientes A y C con ambos métodos:

In [46]:
# Clientes A y C con loc
data.loc[["A","C"]]

level_0  index  customer_id  zip_code            join_date date_of_birth
A        0      0            1     60091  2011-04-17 10:48:33    1994-07-18
C        2      2            3     13244  2011-08-13 15:42:34    2003-11-21

In [47]:
# Clientes A y C con iloc
data.iloc[[0,2]]

level_0  index  customer_id  zip_code            join_date date_of_birth
A        0      0            1     60091  2011-04-17 10:48:33    1994-07-18
C        2      2            3     13244  2011-08-13 15:42:34    2003-11-21

Ahora, de los clientes B y E, queremos su zip_code y su join_date:

In [24]:
# loc


In [25]:
# iloc


Ahora, queremos solo zip_code y join_date, pero de todos los clientes:

In [26]:
# loc


In [27]:
# iloc


¿Qué tal? Ya tenemos varias formas de seleccionar e indexar ciertos datos.

Esto es, sin duda, muy útil. Por otra parte, muchas veces queremos obtener cierta información (clientes, en nuestro ejemplo) que cumplan algunos requisitos. Por ejemplo:
- que sean mayores de 18 años, o
- que su antiguedad en la plataforma sea menor a seis meses, o
- que residan en cierta zona,
- entre otros.

Para ello utilizamos los operadores de comparación (==, >, <, >=, <=, !=).

___
# 3. Filtrado de datos

Los operadores de comparación pueden ser utilizados con Series de pandas. 

Esto resulta ser súper útil para filtrar datos con ciertas condiciones específicas (esto lo veremos enseguida). 

Por ahora, veremos que al usar operadores de comparación con Series, el resultado será Series booleanas. Es decir, cada elemento de dichas series será `True` si la condición se satisface, y `Falso` de lo contrario.

Una vez tenemos la serie boolena, podemos aplicarla para la selección de filas, obteniendo un DataFrame filtrado como resultado.

**Ejemplos**

¿Cuáles clientes son mayores de 18 años?

In [28]:
# Ver tabla


In [29]:
# Adult clients


¿Cuáles clientes se registraron después del primero de Junio del 2011?

In [30]:
# Recent joins


¿Cuáles clientes viven en la zona con CP 13244?

In [31]:
# Zone


Entonces ya tenemos las series booleanas de las que hablábamos. Ahora, ¿cómo podemos usar esto para obtener la información de los clientes que cumplen estas características?

In [32]:
# Filtrar por clientes adultos


In [34]:
# Filtrar por registros recientes


In [35]:
# Se puede hacer todo en una sola línea


In [36]:
# Filtrar por zona


**¡Excelente!**

Los filtrados que acabamos de hacer fueron de una sola condición. Ahora, ¿cómo hacer filtrados con múltiples condiciones?

## 3.1 Filtrado con múltiples condiciones

Cuando queremos filtrar datos con múltiples condiciones, podemos usar operadores booleanos (`and`, `or`, `not`), solo con una pequeña modificación. Usamos:

- `&` en lugar de `and`,
- `|` en lugar de `or`, y
- `~` en lugar de `not`.

**Ejemplos**

Obtener la información de los clientes adultos que vivan en la zona con código postal 13244

In [38]:
# En una sola línea


Obtener la información de los clientes que sean adultos o que se hayan registrado luego del 6 de Junio del 2011

Obtener la información de los clientes que **NO** se hayan registrado en 2011

De esta manera, podemos crear condiciones tan complejas y restrictivas como queramos para filtrar nuestros datos.

Útil, ¿no?

## [Artículo recomendado](https://medium.com/dunder-data/minimally-sufficient-pandas-a8e67f2a2428)

<script>
  $(document).ready(function(){
    $('div.prompt').hide();
    $('div.back-to-top').hide();
    $('nav#menubar').hide();
    $('.breadcrumb').hide();
    $('.hidden-print').hide();
  });
</script>

<footer id="attribution" style="float:right; color:#808080; background:#fff;">
Created with Jupyter by Esteban Jiménez Rodríguez.
</footer>